### PR TITLE
Reorganise QA outputs and write metadata files

### PR DESCRIPTION
**Pull request by KimMcAlpine**
The QA process creates several new products which will be ingested into
the archive. These are:
* a primary beam corrected fits image
* a html QA report
* a fits image of the rms in the primary beam corrected image
* a fits image of the mean background of the primary beam corrected
image

These are created per target in an observation. This PR organises these
outputs so that the primary beam, rms and background images are placed
in separated directories indicated by the suffix '_PB', '_RMS' and
'_BKG' respectively. Each of these directories has a metadata file
describing the archive "FITSImageProduct" in each directory.

Additionally each QA report (per target) is moved to its own directory,
named CBID_continuum_image_TARGET_IClean_PB_continuum_validation_snr5.0_int,
and a metadata file describing the archive "MeerKATReductionProduct" product
in each of these directories is written to that directory.

### CODE REVIEW

**Comment by ctgschollar** [comment]
I would prefer that the FITS Images were separated in the same way as the QA reports. I'd like a 1 to 1 mapping from the QA Report to the data is was generated from. It will simplify my life a lot if I ever need to grab data for a particular QA report. 

I think it makes sense to include the Run number as part of your directory name, I see you separate the QA directories using the target name eg.  1591919459_continuum_image_1607000445862_DEEP_2_QA

What would happen if the same observation imaged the same target twice? Would you end up overwriting the first QA Report and metadata? Or does the second grouping of numbers there disambiguate in those cases? You should be able to include the "Run" value you calculate for the metadata as a way to separate out cases where the same target is imaged twice, as it is based on the index of the scan from katdal (i think), so the Run value will always be different for all images, even if they have the same target.

**Comment by KimMcAlpine** [comment]
> I would prefer that the FITS Images were separated in the same way as the QA reports. I'd like a 1 to 1 mapping from the QA Report to the data is was generated from. It will simplify my life a lot if I ever need to grab data for a particular QA report.
> 
Okay, I will change this as suggested

> I think it makes sense to include the Run number as part of your directory name, I see you separate the QA directories using the target name eg. 1591919459_continuum_image_1607000445862_DEEP_2_QA
> 
> What would happen if the observation imaged the same target twice? Would you end up overwriting the first QA Report and metadata? Or does the second grouping of numbers there disambiguate in those cases? You should be able to include the "Run" value you calculate for the metadata as a way to separate out cases where the same target is imaged twice, as it is based on the index of the scan from katdal (i think), so the Run value will always be different for all images, even if they have the same target.

I can easily include the "Run" number in the directory name, it should be straightforward. 
I think it should be safe to image the same target (or a target with the same name) twice. The directory name has the structure
CBID_continuum_image_START_TIME_TARGET_QA. The start time is the timestamp at the beginning of the pipeline run so that should separate targets from different CBID's. 
For the target names Tom already has a utility function `normalise_target_names` which checks for each new target that the name hasn't already been used in this instance of the pipeline and if it has already been used it appends an index number to the end of the name  e.g if you imaged 'DEEP_2' twice then the second in the second image the target would be named 'DEEP_2_1'.  As I'm reusing the metadata and image names from Tom's original continuum images it should guarantee that the names are unique per target as well.

File: katacomb/katacomb/qa_report.py [hunk 1]
```diff
@@ -60,40 +65,188 @@ def make_pbeam_images(metadata, in_dir, out_dir):
...
+    meta_suffix['Description'] = desc_prefix + ': Continuum image PB corrected'
+
+    write_metadata(meta_suffix, outdir)
+
+
+def make_report_metadata(metadata, out_dir):
+    """Write a new reduction product metadata file.
+
+    It is based on the original pipeline metadata, but with
+    updated 'Description' and 'ReductionName' keys.
```
Comment (ludwigschwardt) [inline katacomb/katacomb/qa_report.py:207]: Empty line needed above `Parameters`.

File: katacomb/katacomb/qa_report.py [hunk 2]
```diff
@@ -60,40 +65,188 @@ def make_pbeam_images(metadata, in_dir, out_dir):
...
 
+def organise_qa_output(metadata, pb_dir, qa_dir, rms_dir, bkg_dir):
+    """Organise QA output into separate directories.
+
+    Move the rms and mean images created by the QA report code to
+    dedicated per image type directories. Create metadata files per
+    image type directory. Move each QA report directory to
+    a given directory, write a metadata file in each individual report
+    directory.
+    Parameters
```
Comment (ludwigschwardt) [inline katacomb/katacomb/qa_report.py:113]: Empty line needed above `Parameters`.

**Review by ludwigschwardt — CHANGES_REQUESTED** [review]
Generally looks good!

I have one style suggestion (maybe just a personal habit 😄 ): I usually put function/class definitions above the first line that uses them, so when I see e.g. `_caption_pngs(...)` I tend to scroll up to find the definition. The definitions in this file are in no particular order, which means more scrolling for me.

File: katacomb/katacomb/qa_report.py (hunk 1 above)
Comment (KimMcAlpine) [reply katacomb/katacomb/qa_report.py:202]: > Generally looks good!
> 
> I have one style suggestion (maybe just a personal habit smile ): I usually put function/class definitions above the first line that uses them, so when I see e.g. `_caption_pngs(...)` I tend to scroll up to find the definition. The definitions in this file are in no particular order, which means more scrolling for me.

I've made an attempt to re-arrange the functions according to this principle.

File: katacomb/katacomb/qa_report.py (hunk 2 above)
Comment (KimMcAlpine) [reply katacomb/katacomb/qa_report.py]: Done

Comment (KimMcAlpine) [reply katacomb/katacomb/qa_report.py]: Done

**Review by ludwigschwardt — APPROVED** [review]

**Comment by ctgschollar** [comment]
I'm happy with the metadata, but I found that there is a 1591919459_continuum_image_1607000445862_DEEP_2_2_RMS.writing directory inside the 1591919459_continuum_image_1607000445862_DEEP_2_2_RMS directory for the example data you gave me. With a copy of the fits file in it. Not sure if that is something that happened because you ran multiple tests of the pipeline. It doesn't seem to have happened with any of the other directories

**Review by ctgschollar — APPROVED** [review]
Managed to ingest some example data. I have updates for the transfer side, so we can deploy this to site, will just have to co-ordinate updating the transfer side and the generation side at the same time.

**Pull request merged**